packages/api-server/src/methods/validator.ts

```diff
@@ -129,316 +111,417 @@ export const validators = {
-    if ((!validateHexString(topic) || topic.length !== 66) && topic !== null) {
-      return invalidParamsError(index, `invalid argument`);
+export function verifyNewFilterObj(
+  filterObj: any,
+  index: number
+): InvalidParamsError | undefined {
+  if (typeof filterObj !== "object") {
+    return invalidParamsError(index, `filter argument must be an object`);
+  }
+
+  const fromBlock = filterObj.fromBlock;
+  const toBlock = filterObj.toBlock;
+  const address = filterObj.address;
+  const topics = filterObj.topics;
+
+  // validate `fromBlock`
+  if (fromBlock != null) {
+    const fromBlockErr = verifyBlockParameter(fromBlock, index);
+    if (fromBlockErr) {
+      return padErrorContext(fromBlockErr, "filter fromBlock");
     }
   }
 
-  if (Array.isArray(topic)) {
-    for (const t of topic) {
-      if ((!validateHexString(t) || t.length !== 66) && t !== null) {
-        return invalidParamsError(index, `invalid argument`);
+  // validate `toBlock`
+  if (toBlock != null) {
+    const toBlockErr = verifyBlockParameter(toBlock, index);
+    if (toBlockErr) {
+      return padErrorContext(toBlockErr, "filter toBlock");
+    }
+  }
+
+  // validate `address`
+  if (address != null) {
+    if (Array.isArray(address)) {
+      for (const addr of address) {
+        const addressErr = verifyAddress(addr, index);
+        if (addressErr) {
+          return padErrorContext(addressErr, "filter address[] Array");
+        }
+      }
+    } else {
+      const addressErr = verifyAddress(address, index);
+      if (addressErr) {
+        return padErrorContext(addressErr, "filter address");
+      }
+    }
+  }
+
+  // validate `topics`
+  if (topics != null) {
+    if (!Array.isArray(topics)) {
+      return invalidParamsError(index, `filter topics must be an array`);
+    }
+    for (const topic of topics) {
+      const topicErr = verifyFilterTopic(topic, index);
+      if (topicErr) {
+        return padErrorContext(topicErr, "filter topic[] Array");
       }
     }
   }
 
   return undefined;
 }
 
-function validateAddress(address: string): boolean {
-  return /^0x[0-9a-fA-F]+$/.test(address) && address.length === 42;
+export function verifyGasLimit(
+  gasLimit: HexString,
+  index: number
+): InvalidParamsError | undefined {
+  const gasLimitErr = verifyHexNumber(gasLimit, index);
+  if (gasLimitErr) {
+    return padErrorContext(gasLimitErr, "gasLimit");
+  }
+
+  if (BigInt(gasLimit) > BigInt(POLY_MAX_BLOCK_GAS_LIMIT)) {
+    return invalidParamsError(
+      index,
+      `gas limit ${gasLimit} exceeds block gas limit of ${POLY_MAX_BLOCK_GAS_LIMIT}`
+    );
+  }
+  return undefined;
 }
+//******* end of standalone verify function ********/
 
-function invalidParamsError(index: number, message: string): any {
+// some utils function
+function invalidParamsError(index: number, message: string) {
   return new InvalidParamsError(`invalid argument ${index}: ${message}`);
 }
+
+export function padErrorContext(
```

Comment:
How about add a new method like `InvalidParamsError#addContext` as a class method ?

```typescript
class InvalidParamsError {
  public addContext(context: string): InvalidParamsError {
    this.message = ...;
    return this;
  }
}
```

When use 

```typescript
  if (gasLimitErr) {
    return padErrorContext(gasLimitErr, "gasLimit");
  }
```

Change to:
```typescript
return gasLimitErr.addContext("gasLimit");
```

Reply:
sounds better. will update.

Reply:
updated